### ISSUE TITLE
CODEOWNERS global approvers plan

### ISSUE THREAD

**Issue by alxhub** [angular/angular#28245]
# Problem

`CODEOWNERS` maps each file to a group of approvers. Unfortunately, it can only have one group allowed per file - we cannot have groups of "global approvers" with the ability to approve changes across larger parts of the project.

We would like this capability as otherwise the review burden on large changes is too great. We cannot add individual users to every group as this creates review spam.

# Proposal

We already have a bot that monitors PR labels, posts PR statuses, etc. We will add this bot's user to every approval group in `CODEOWNERS`, making it a _de facto_ global approver.

The bot will be programmed with a configuration of global approvers and a list of review groups (e.g. `fw-core`) for which they are considered as global approvers.

For each PR, the bot will take action when all of the subsequent conditions are met:

1) A configured global approver (e.g. @kara) has approved the PR.

2) The same configured global approver has added the `PR state: global approval` label.

3) The PR has all _other_ approvals it needs (those not covered by the global approver(s).

When all 3 of these conditions are met, the bot itself will approve the PR. Because of its status as an approver for every group, this will enable the PR to merge.

**Comment by IgorMinar** [comment]
This sounds reasonable to me. I'm not sure why we need the third check though. If the PR is still missing some approval for a pattern not covered by the global approval group (e.g. fw-public-api and fw-dev-infra paths are not in the global approver's list to prevent accidents), the github should not mark the PR as green because it's missing approvals regardless of the approval added by the bot.

@ocombe is this doable? @alxhub pointed out that approvals can now be granted via the api as well: https://developer.github.com/v3/pulls/reviews/#submit-a-pull-request-review

**Comment by ocombe** [comment]
This seems doable, but keep in mind that:
- it gives a lot of powers to the bot and to me (since I code the bot), is this ok with you? I don't expect this to be an issue, since it's not merging the PR, just approving it, but I prefer to be upfront about it
- the api is sometimes unstable (either because of bugs in the bot, or because of github not sending events, or because of rate limitations, or because of firebase issues)

**Comment by IgorMinar** [comment]
In fine with power delegation as long as we review how secrets are stored
and manipulated with.

When it comes to unreliability - could we solved this more generally via
queueing operations and retrying when something goes wrong, and dequeuing
only operations that we confirmed were executed successfully?

**Comment by ocombe** [comment]
yes, but we need to be careful with this to avoid loops (where it tries again and again to make the request pass) in case the error is legit and not due to a temporary problem, because we pay firebase and we could end up with a gigantic bill (I've heard such stories).
